using System;
using CliWrap.Exceptions;
using CliWrap.Internal;

namespace CliWrap.Models
{
    /// <summary>
    /// Output generated by executing a command line interface
    /// </summary>
    public class ExecutionOutput
    {
        /// <summary>
        /// Exit code
        /// </summary>
        public int ExitCode { get; }

        /// <summary>
        /// Standard output
        /// </summary>
        public string StandardOutput { get; }

        /// <summary>
        /// Standard error
        /// </summary>
        public string StandardError { get; }

        /// <summary>
        /// Whether there is an error
        /// </summary>
        public bool HasError => !string.IsNullOrEmpty(StandardError);

        /// <summary>
        /// When the execution started
        /// </summary>
        public DateTime StartTime { get; }

        /// <summary>
        /// When the execution finished
        /// </summary>
        public DateTime ExitTime { get; }

        /// <summary>
        /// Execution run time
        /// </summary>
        public TimeSpan RunTime => ExitTime - StartTime;

        /// <summary />
        public ExecutionOutput(int exitCode, string standardOutput, string standardError,
            DateTime startTime, DateTime exitTime)
        {
            ExitCode = exitCode;
            StandardOutput = standardOutput.GuardNotNull(nameof(standardOutput));
            StandardError = standardError.GuardNotNull(nameof(standardError));
            StartTime = startTime;
            ExitTime = exitTime;
        }

        /// <summary>
        /// Throws a <see cref="StandardErrorException"/> if there is an error associated with this <see cref="ExecutionOutput"/>
        /// </summary>
        public void ThrowIfError()
        {
            if (HasError)
                throw new StandardErrorException(StandardError);
        }
    }
}